File: sdk/typescript/src/proof.ts
import { createHash, randomBytes } from "crypto";

/**
 * Client-side Proof-of-Human generation toolkit.
 *
 * Generates behavioral fingerprints and challenge responses for
 * the your.fun verification pipeline. The fingerprint captures
 * interaction timing, input entropy, and cognitive response patterns.
 */

/**
 * Raw behavioral event captured from user interactions.
 */
interface BehavioralEvent {
    type: "keystroke" | "mouse" | "scroll" | "focus" | "idle";
    timestamp: number;
    data: Record<string, number>;
}

/**
 * Aggregated behavioral metrics from a collection of events.
 */
interface BehavioralMetrics {
    keystrokeTimingVariance: number;
    mouseMovementEntropy: number;
    scrollPatternScore: number;
    focusSwitchFrequency: number;
    idlePatternScore: number;
    totalEvents: number;
    sessionDurationMs: number;
}

/**
 * Final fingerprint output for on-chain registration.
 */
interface FingerprintOutput {
    hash: Uint8Array;
    metrics: BehavioralMetrics;
    confidence: number;
}

/**
 * Collects and analyzes behavioral data to generate a Proof-of-Human fingerprint.
 */
export class ProofGenerator {
    private events: BehavioralEvent[] = [];
    private sessionStartTime: number;
    private readonly minEventsRequired = 50;
    private readonly minSessionDurationMs = 5000;

    constructor() {
        this.sessionStartTime = Date.now();
    }

    /**
     * Records a keystroke timing event.
     * Captures the interval between keypresses and the hold duration.
     */
    recordKeystroke(keyDownTime: number, keyUpTime: number): void {
        const holdDuration = keyUpTime - keyDownTime;
        const interval = this.events.length > 0
            ? keyDownTime - this.events[this.events.length - 1].timestamp
            : 0;

        this.events.push({
            type: "keystroke",
            timestamp: keyDownTime,
            data: {
                holdDuration,
                interval,
                keyUpTime,
            },
        });
    }

    /**
     * Records a mouse movement event.
     * Captures position, velocity, and acceleration vectors.
     */
    recordMouseMovement(x: number, y: number, timestamp: number): void {
        const lastMouse = this.getLastEventOfType("mouse");
        let velocity = 0;
        let acceleration = 0;

        if (lastMouse) {
            const dt = timestamp - lastMouse.timestamp;
            if (dt > 0) {
                const dx = x - lastMouse.data.x;
                const dy = y - lastMouse.data.y;
                velocity = Math.sqrt(dx * dx + dy * dy) / dt;

                if (lastMouse.data.velocity > 0) {
                    acceleration = (velocity - lastMouse.data.velocity) / dt;
                }
            }
        }

        this.events.push({
            type: "mouse",
            timestamp,
            data: { x, y, velocity, acceleration },
        });
    }

    /**
     * Records a scroll event with direction and intensity.
     */
    recordScroll(deltaY: number, timestamp: number): void {
        this.events.push({
            type: "scroll",
            timestamp,
            data: {
                deltaY,
                intensity: Math.abs(deltaY),
            },
        });
    }

    /**
     * Records a focus change event (tab switching, window focus).
     */
    recordFocusChange(hasFocus: boolean, timestamp: number): void {
        this.events.push({
            type: "focus",
            timestamp,
            data: {
                hasFocus: hasFocus ? 1 : 0,
            },
        });
    }

    /**
     * Records an idle state detection based on inactivity threshold.
     */
    recordIdlePeriod(startTime: number, endTime: number): void {
        this.events.push({
            type: "idle",
            timestamp: startTime,
            data: {
                duration: endTime - startTime,
            },
        });
    }

    /**
     * Computes aggregated behavioral metrics from recorded events.
     */
    computeMetrics(): BehavioralMetrics {
        const sessionDurationMs = Date.now() - this.sessionStartTime;

        const keystrokes = this.events.filter((e) => e.type === "keystroke");
        const keystrokeTimingVariance = this.computeVariance(
            keystrokes.map((e) => e.data.interval).filter((v) => v > 0)
        );

        const mouseEvents = this.events.filter((e) => e.type === "mouse");
        const mouseMovementEntropy = this.computeEntropy(
            mouseEvents.map((e) => e.data.velocity)
        );

        const scrollEvents = this.events.filter((e) => e.type === "scroll");
        const scrollPatternScore = this.computeScrollPattern(scrollEvents);

        const focusEvents = this.events.filter((e) => e.type === "focus");
        const focusSwitchFrequency = focusEvents.length / (sessionDurationMs / 1000);

        const idleEvents = this.events.filter((e) => e.type === "idle");
        const idlePatternScore = this.computeIdlePattern(idleEvents, sessionDurationMs);

        return {
            keystrokeTimingVariance,
            mouseMovementEntropy,
            scrollPatternScore,
            focusSwitchFrequency,
            idlePatternScore,
            totalEvents: this.events.length,
            sessionDurationMs,
        };
    }

    /**
     * Generates the final fingerprint hash from behavioral metrics.
     * Returns a 32-byte hash suitable for on-chain storage.
     */
    generateFingerprint(): FingerprintOutput {
        const metrics = this.computeMetrics();

        const dataBuffer = Buffer.alloc(56);
        dataBuffer.writeDoubleBE(metrics.keystrokeTimingVariance, 0);
        dataBuffer.writeDoubleBE(metrics.mouseMovementEntropy, 8);
        dataBuffer.writeDoubleBE(metrics.scrollPatternScore, 16);
        dataBuffer.writeDoubleBE(metrics.focusSwitchFrequency, 24);
        dataBuffer.writeDoubleBE(metrics.idlePatternScore, 32);
        dataBuffer.writeUInt32BE(metrics.totalEvents, 40);
        dataBuffer.writeDoubleBE(metrics.sessionDurationMs, 44);

        const hash = createHash("sha256").update(dataBuffer).digest();
        const fingerprintHash = new Uint8Array(hash);
        const confidence = this.computeConfidence(metrics);

        return {
            hash: fingerprintHash,
            metrics,
            confidence,
        };
    }

    /**
     * Generates a challenge nonce for the verification process.
     */
    generateChallengeNonce(): Uint8Array {
        return new Uint8Array(randomBytes(32));
    }

    /**
     * Solves a challenge by combining the nonce with the fingerprint hash.
     * Must match the on-chain compute_challenge_hash function.
     */
    solveChallenge(nonce: Uint8Array, fingerprintHash: Uint8Array): Uint8Array {
        const result = new Uint8Array(32);

        for (let i = 0; i < 32; i++) {
            result[i] = nonce[i] ^ fingerprintHash[i];
            result[i] = (result[i] + nonce[(i + 7) % 32]) & 0xff;
            result[i] ^= fingerprintHash[(i + 13) % 32];
        }

        for (let round = 0; round < 4; round++) {
            for (let i = 0; i < 32; i++) {
                const prev = result[(i + 31) % 32];
                const next = result[(i + 1) % 32];
                result[i] = (result[i] + ((prev * next) & 0xff) + round) & 0xff;
            }
        }

        return result;
    }

    /**
     * Returns whether enough data has been collected for verification.
     */
    isReadyForVerification(): boolean {
        const sessionDuration = Date.now() - this.sessionStartTime;
        return (
            this.events.length >= this.minEventsRequired &&
            sessionDuration >= this.minSessionDurationMs
        );
    }

    /**
     * Returns the number of events currently collected.
     */
    getEventCount(): number {
        return this.events.length;
    }

    /**
     * Resets all collected behavioral data.
     */
    reset(): void {
        this.events = [];
        this.sessionStartTime = Date.now();
    }

    // -- Private computational methods --

    private computeVariance(values: number[]): number {
        if (values.length < 2) return 0;
        const mean = values.reduce((a, b) => a + b, 0) / values.length;
        const squaredDiffs = values.map((v) => (v - mean) ** 2);
        return squaredDiffs.reduce((a, b) => a + b, 0) / (values.length - 1);
    }

    private computeEntropy(values: number[]): number {
        if (values.length < 2) return 0;

        const binCount = 20;
        const min = Math.min(...values);
        const max = Math.max(...values);
        const range = max - min || 1;
        const bins = new Array(binCount).fill(0);

        for (const value of values) {
            const binIndex = Math.min(
                Math.floor(((value - min) / range) * binCount),
                binCount - 1
            );
            bins[binIndex]++;
        }

        let entropy = 0;
        for (const count of bins) {
            if (count > 0) {
                const p = count / values.length;
                entropy -= p * Math.log2(p);
            }
        }

        return entropy;
    }

    private computeScrollPattern(scrollEvents: BehavioralEvent[]): number {
        if (scrollEvents.length < 3) return 0;

        let directionChanges = 0;
        let prevDirection = 0;

        for (const event of scrollEvents) {
            const direction = event.data.deltaY > 0 ? 1 : -1;
            if (prevDirection !== 0 && direction !== prevDirection) {
                directionChanges++;
            }
            prevDirection = direction;
        }

        const changeRatio = directionChanges / scrollEvents.length;
        const intensities = scrollEvents.map((e) => e.data.intensity);
        const intensityVariance = this.computeVariance(intensities);

        return changeRatio * 50 + Math.min(intensityVariance / 100, 50);
    }

    private computeIdlePattern(
        idleEvents: BehavioralEvent[],
        sessionDurationMs: number
    ): number {
        if (idleEvents.length === 0) return 100;

        const totalIdleTime = idleEvents.reduce((sum, e) => sum + e.data.duration, 0);
        const idleRatio = totalIdleTime / sessionDurationMs;

        const idleDurations = idleEvents.map((e) => e.data.duration);
        const durationVariance = this.computeVariance(idleDurations);

        const baseScore = (1 - idleRatio) * 70;
        const varianceBonus = Math.min(durationVariance / 10000, 30);

        return baseScore + varianceBonus;
    }

    private computeConfidence(metrics: BehavioralMetrics): number {
        let score = 0;

        if (metrics.totalEvents >= 100) score += 20;
        else if (metrics.totalEvents >= 50) score += 10;

        if (metrics.sessionDurationMs >= 30000) score += 20;
        else if (metrics.sessionDurationMs >= 10000) score += 10;

        if (metrics.keystrokeTimingVariance > 1000) score += 15;
        if (metrics.mouseMovementEntropy > 2.0) score += 15;
        if (metrics.scrollPatternScore > 30) score += 10;
        if (metrics.focusSwitchFrequency > 0.05 && metrics.focusSwitchFrequency < 2.0) score += 10;
        if (metrics.idlePatternScore > 50) score += 10;

        return Math.min(score, 100);
    }

    private getLastEventOfType(type: string): BehavioralEvent | undefined {
        for (let i = this.events.length - 1; i >= 0; i--) {
            if (this.events[i].type === type) return this.events[i];
        }
        return undefined;
    }
}

File: sdk-python/yourfun/proof.py
"""
Proof-of-Human fingerprint generator for the your.fun platform.
Collects behavioral interaction data and produces cryptographic fingerprints
suitable for on-chain verification.
"""

from __future__ import annotations
import hashlib
import os
import math
import time
from dataclasses import dataclass, field
from typing import Optional
import struct


@dataclass
class BehavioralEvent:
    """A single captured behavioral event."""
    event_type: str
    timestamp: float
    data: dict[str, float]


@dataclass
class BehavioralMetrics:
    """Aggregated metrics from behavioral event analysis."""
    keystroke_timing_variance: float
    mouse_movement_entropy: float
    scroll_pattern_score: float
    focus_switch_frequency: float
    idle_pattern_score: float
    total_events: int
    session_duration_ms: float


@dataclass
class FingerprintResult:
    """Generated fingerprint output for on-chain registration."""
    hash_bytes: bytes
    metrics: BehavioralMetrics
    confidence: float


class ProofGenerator:
    """
    Collects and analyzes behavioral data to generate a Proof-of-Human fingerprint.

    Usage:
        generator = ProofGenerator()
        generator.record_keystroke(down_time, up_time)
        generator.record_mouse_movement(x, y, timestamp)

        if generator.is_ready():
            result = generator.generate_fingerprint()
            nonce = generator.generate_challenge_nonce()
            solution = generator.solve_challenge(nonce, result.hash_bytes)
    """

    MIN_EVENTS_REQUIRED = 50
    MIN_SESSION_DURATION_MS = 5000.0

    def __init__(self) -> None:
        self._events: list[BehavioralEvent] = []
        self._session_start: float = time.time() * 1000

    def record_keystroke(self, key_down_time: float, key_up_time: float) -> None:
        """Records a keystroke timing event with hold duration and interval."""
        hold_duration = key_up_time - key_down_time
        interval = 0.0
        if self._events:
            interval = key_down_time - self._events[-1].timestamp

        self._events.append(BehavioralEvent(
            event_type="keystroke",
            timestamp=key_down_time,
            data={
                "hold_duration": hold_duration,
                "interval": interval,
                "key_up_time": key_up_time,
            },
        ))

    def record_mouse_movement(self, x: float, y: float, timestamp: float) -> None:
        """Records a mouse movement with computed velocity and acceleration."""
        velocity = 0.0
        acceleration = 0.0

        last_mouse = self._get_last_event_of_type("mouse")
        if last_mouse:
            dt = timestamp - last_mouse.timestamp
            if dt > 0:
                dx = x - last_mouse.data.get("x", 0)
                dy = y - last_mouse.data.get("y", 0)
                velocity = math.sqrt(dx * dx + dy * dy) / dt
                prev_velocity = last_mouse.data.get("velocity", 0)
                if prev_velocity > 0:
                    acceleration = (velocity - prev_velocity) / dt

        self._events.append(BehavioralEvent(
            event_type="mouse",
            timestamp=timestamp,
            data={"x": x, "y": y, "velocity": velocity, "acceleration": acceleration},
        ))

    def record_scroll(self, delta_y: float, timestamp: float) -> None:
        """Records a scroll event with direction and intensity."""
        self._events.append(BehavioralEvent(
            event_type="scroll",
            timestamp=timestamp,
            data={"delta_y": delta_y, "intensity": abs(delta_y)},
        ))

    def record_focus_change(self, has_focus: bool, timestamp: float) -> None:
        """Records a window focus change event."""
        self._events.append(BehavioralEvent(
            event_type="focus",
            timestamp=timestamp,
            data={"has_focus": 1.0 if has_focus else 0.0},
        ))

    def record_idle_period(self, start_time: float, end_time: float) -> None:
        """Records a detected idle period."""
        self._events.append(BehavioralEvent(
            event_type="idle",
            timestamp=start_time,
            data={"duration": end_time - start_time},
        ))

    def compute_metrics(self) -> BehavioralMetrics:
        """Computes aggregated behavioral metrics from all recorded events."""
        session_duration_ms = time.time() * 1000 - self._session_start

        keystrokes = [e for e in self._events if e.event_type == "keystroke"]
        intervals = [e.data["interval"] for e in keystrokes if e.data["interval"] > 0]
        keystroke_variance = self._compute_variance(intervals)

        mouse_events = [e for e in self._events if e.event_type == "mouse"]
        velocities = [e.data["velocity"] for e in mouse_events]
        mouse_entropy = self._compute_entropy(velocities)

        scroll_events = [e for e in self._events if e.event_type == "scroll"]
        scroll_score = self._compute_scroll_pattern(scroll_events)

        focus_events = [e for e in self._events if e.event_type == "focus"]
        focus_freq = len(focus_events) / max(session_duration_ms / 1000, 0.001)

        idle_events = [e for e in self._events if e.event_type == "idle"]
        idle_score = self._compute_idle_pattern(idle_events, session_duration_ms)

        return BehavioralMetrics(
            keystroke_timing_variance=keystroke_variance,
            mouse_movement_entropy=mouse_entropy,
            scroll_pattern_score=scroll_score,
            focus_switch_frequency=focus_freq,
            idle_pattern_score=idle_score,
            total_events=len(self._events),
            session_duration_ms=session_duration_ms,
        )

    def generate_fingerprint(self) -> FingerprintResult:
        """Generates a 32-byte SHA-256 fingerprint hash from behavioral metrics."""
        metrics = self.compute_metrics()

        data_buffer = struct.pack(
            ">dddddId",
            metrics.keystroke_timing_variance,
            metrics.mouse_movement_entropy,
            metrics.scroll_pattern_score,
            metrics.focus_switch_frequency,
            metrics.idle_pattern_score,
            metrics.total_events,
            metrics.session_duration_ms,
        )

        hash_bytes = hashlib.sha256(data_buffer).digest()
        confidence = self._compute_confidence(metrics)

        return FingerprintResult(
            hash_bytes=hash_bytes,
            metrics=metrics,
            confidence=confidence,
        )

    def generate_challenge_nonce(self) -> bytes:
        """Generates a cryptographically secure 32-byte challenge nonce."""
        return os.urandom(32)

    def solve_challenge(self, nonce: bytes, fingerprint_hash: bytes) -> bytes:
        """
        Solves a verification challenge using the nonce and fingerprint hash.
        This must produce output identical to the on-chain compute_challenge_hash.
        """
        result = bytearray(32)

        for i in range(32):
            result[i] = nonce[i] ^ fingerprint_hash[i]
            result[i] = (result[i] + nonce[(i + 7) % 32]) & 0xFF
            result[i] ^= fingerprint_hash[(i + 13) % 32]

        for round_num in range(4):
            for i in range(32):
                prev = result[(i + 31) % 32]
                nxt = result[(i + 1) % 32]
                result[i] = (result[i] + ((prev * nxt) & 0xFF) + round_num) & 0xFF

        return bytes(result)

    def is_ready(self) -> bool:
        """Returns whether enough data has been collected for verification."""
        session_duration = time.time() * 1000 - self._session_start
        return (
            len(self._events) >= self.MIN_EVENTS_REQUIRED
            and session_duration >= self.MIN_SESSION_DURATION_MS
        )

    @property
    def event_count(self) -> int:
        return len(self._events)

    def reset(self) -> None:
        """Resets all collected behavioral data."""
        self._events.clear()
        self._session_start = time.time() * 1000

    # -- Private computational methods --

    @staticmethod
    def _compute_variance(values: list[float]) -> float:
        if len(values) < 2:
            return 0.0
        mean = sum(values) / len(values)
        return sum((v - mean) ** 2 for v in values) / (len(values) - 1)

    @staticmethod
    def _compute_entropy(values: list[float]) -> float:
        if len(values) < 2:
            return 0.0

        bin_count = 20
        min_val = min(values)
        max_val = max(values)
        range_val = max_val - min_val if max_val != min_val else 1.0
        bins = [0] * bin_count

        for value in values:
            bin_idx = min(int(((value - min_val) / range_val) * bin_count), bin_count - 1)
            bins[bin_idx] += 1

        entropy = 0.0
        n = len(values)
        for count in bins:
            if count > 0:
                p = count / n
                entropy -= p * math.log2(p)

        return entropy

    @staticmethod
    def _compute_scroll_pattern(scroll_events: list[BehavioralEvent]) -> float:
        if len(scroll_events) < 3:
            return 0.0

        direction_changes = 0
        prev_direction = 0

        for event in scroll_events:
            direction = 1 if event.data["delta_y"] > 0 else -1
            if prev_direction != 0 and direction != prev_direction:
                direction_changes += 1
            prev_direction = direction

        change_ratio = direction_changes / len(scroll_events)
        intensities = [e.data["intensity"] for e in scroll_events]
        intensity_variance = ProofGenerator._compute_variance(intensities)

        return change_ratio * 50 + min(intensity_variance / 100, 50)

    @staticmethod
    def _compute_idle_pattern(
        idle_events: list[BehavioralEvent], session_duration_ms: float
    ) -> float:
        if not idle_events:
            return 100.0

        total_idle = sum(e.data["duration"] for e in idle_events)
        idle_ratio = total_idle / max(session_duration_ms, 1.0)

        durations = [e.data["duration"] for e in idle_events]
        duration_variance = ProofGenerator._compute_variance(durations)

        base_score = (1 - idle_ratio) * 70
        variance_bonus = min(duration_variance / 10000, 30)

        return base_score + variance_bonus

    @staticmethod
    def _compute_confidence(metrics: BehavioralMetrics) -> float:
        score = 0.0

        if metrics.total_events >= 100:
            score += 20
        elif metrics.total_events >= 50:
            score += 10

        if metrics.session_duration_ms >= 30000:
            score += 20
        elif metrics.session_duration_ms >= 10000:
            score += 10

        if metrics.keystroke_timing_variance > 1000:
            score += 15
        if metrics.mouse_movement_entropy > 2.0:
            score += 15
        if metrics.scroll_pattern_score > 30:
            score += 10
        if 0.05 < metrics.focus_switch_frequency < 2.0:
            score += 10
        if metrics.idle_pattern_score > 50:
            score += 10

        return min(score, 100.0)

    def _get_last_event_of_type(self, event_type: str) -> Optional[BehavioralEvent]:
        for event in reversed(self._events):
            if event.event_type == event_type:
                return event
        return None
